### ISSUE TITLE
Flaky crashes on service/async_generator_breakpoint_test/1

### ISSUE THREAD

**Issue by alexmarkov** [dart-lang/sdk#38108]
In debug mode there is an assertion failure `error: expected: !function.ForceOptimize()`:

```
/===============================================================================\
  | service/async_generator_breakpoint_test/1 failed again (Crash, expected Pass) |
  \===============================================================================/
  --- Command "vm" (took 21.000824s):
  DART_CONFIGURATION=DebugX64 out/DebugX64/dart --verbose-debug --stacktrace-every=55 --stress-async-stacks --ignore-unrecognized-flags --packages=/b/s/w/ir/.packages /b/s/w/ir/runtime/observatory/tests/service/async_generator_breakpoint_test.dart
  exit code:
  -6
  stdout:
  ** Launching SERVICE_TEST_TESTEE=true out/DebugX64/dart --disable-service-auth-codes --profiler --verbose-debug --stacktrace-every=55 --stress-async-stacks --ignore-unrecognized-flags --packages=/b/s/w/ir/.packages --enable-vm-service:0 /b/s/w/ir/runtime/observatory/tests/service/async_generator_breakpoint_test.dart
  >testee>out> Observatory listening on http://127.0.0.1:37919/
  ** Signaled to run test queries on http://127.0.0.1:37919/
  >testee>out>
  Loading VM...
  >testee>out> testee waiting
  Done loading VM
  Running async_generator_breakpoint_test.dart [1/1]
  >testee>err> Registering pending breakpoint for uncompiled function 'file:///b/s/w/ir/runtime/observatory/tests/service/async_generator_breakpoint_test.dart_::_printSync' at line 13 col 1
  >testee>err> Registering pending breakpoint for uncompiled function 'file:///b/s/w/ir/runtime/observatory/tests/service/async_generator_breakpoint_test.dart_::_printAsync' at line 18 col 1
  >testee>err> Registering pending breakpoint for uncompiled function 'file:///b/s/w/ir/runtime/observatory/tests/service/async_generator_breakpoint_test.dart_::_printAsyncStar' at line 23 col 1
  >testee>err> Registering pending breakpoint for uncompiled function 'file:///b/s/w/ir/runtime/observatory/tests/service/async_generator_breakpoint_test.dart_::_printSyncStar' at line 27 col 1
  >testee>err> Setting code breakpoint at pos 1138 pc 0x7fd1e803c7c0 offset 0x160
  >testee>err> Resolved code breakpoint for function 'file:///b/s/w/ir/runtime/observatory/tests/service/async_generator_breakpoint_test.dart_::_testeeDo' at line 44 col 3
  stderr:
  ../../runtime/vm/compiler/jit/compiler.cc: 919: error: expected: !function.ForceOptimize()
  version=2.6.0-edge.1f9bf308b23e7bd18589dfd80295a90905ea07b8 (Thu Aug 29 15:49:59 2019 +0000) on "linux_x64"
  thread=4569, isolate=main(0x55b14f610c00)
    pc 0x000055b14d8a4a8c fp 0x00007ff8880fd4a0 dart::Profiler::DumpStackTrace(void*)
    pc 0x000055b14d4bbb42 fp 0x00007ff8880fd580 dart::Assert::Fail(char const*, ...)
    pc 0x000055b14db1e4ff fp 0x00007ff8880fd5d0 dart::Compiler::EnsureUnoptimizedCode(dart::Thread*, dart::Function const&)
    pc 0x000055b14d73c63c fp 0x00007ff8880fd630 dart::DeferredPp::Materialize(dart::DeoptContext*)
    pc 0x000055b14d73eea0 fp 0x00007ff8880fd7b0 dart::DeoptContext::MaterializeDeferredObjects()
    pc 0x000055b14d7319b6 fp 0x00007ff8880fd800 dart::Debugger::DeoptimizeToArray(dart::Thread*, dart::StackFrame*, dart::Code const&)
    pc 0x000055b14d72648d fp 0x00007ff8880fda00 dart::Debugger::CollectAwaiterReturnStackTrace()
    pc 0x000055b14d917a7b fp 0x00007ff8880fdc10 dart::DRT_StackOverflow(dart::NativeArguments)
    pc 0x00007ff888601197 fp 0x00007ff8880fdc50 Unknown symbol
    pc 0x00007ff882e7ae81 fp 0x00007ff8880fdc70 Unknown symbol
    pc 0x00007ff882e7a4a8 fp 0x00007ff8880fdce8 Unknown symbol
    pc 0x00007ff882e7a091 fp 0x00007ff8880fdd58 Unknown symbol
    pc 0x00007ff882a94ca3 fp 0x00007ff8880fdde8 Unknown symbol
    pc 0x00007ff882e8febc fp 0x00007ff8880fde50 Unknown symbol
    pc 0x00007ff882e8fbe0 fp 0x00007ff8880fdea0 Unknown symbol
    pc 0x00007ff882e81bf5 fp 0x00007ff8880fdf20 Unknown symbol
    pc 0x00007ff882e819d4 fp 0x00007ff8880fdf98 Unknown symbol
    pc 0x00007ff882e81224 fp 0x00007ff8880fe040 Unknown symbol
    pc 0x00007ff882e8d86c fp 0x00007ff8880fe0d0 Unknown symbol
    pc 0x00007ff882e8d4f2 fp 0x00007ff8880fe128 Unknown symbol
    pc 0x00007ff882e8d012 fp 0x00007ff8880fe168 Unknown symbol
    pc 0x00007ff882e8cc83 fp 0x00007ff8880fe1a8 Unknown symbol
    pc 0x00007ff882e8c8e0 fp 0x00007ff8880fe1e8 Unknown symbol
    pc 0x00007ff882e8c383 fp 0x00007ff8880fe228 Unknown symbol
    pc 0x00007ff882e8dd06 fp 0x00007ff8880fe278 Unknown symbol
    pc 0x00007ff882e8dae0 fp 0x00007ff8880fe2c8 Unknown symbol
    pc 0x00007ff882e81c80 fp 0x00007ff8880fe348 Unknown symbol
    pc 0x00007ff882e819d4 fp 0x00007ff8880fe3c0 Unknown symbol
    pc 0x00007ff882e81224 fp 0x00007ff8880fe468 Unknown symbol
    pc 0x00007ff882e8d86c fp 0x00007ff8880fe4f8 Unknown symbol
    pc 0x00007ff882e8d4f2 fp 0x00007ff8880fe550 Unknown symbol
    pc 0x00007ff882e8d012 fp 0x00007ff8880fe590 Unknown symbol
    pc 0x00007ff882e8cc83 fp 0x00007ff8880fe5d0 Unknown symbol
    pc 0x00007ff882e8c8e0 fp 0x00007ff8880fe610 Unknown symbol
    pc 0x00007ff882e8c383 fp 0x00007ff8880fe650 Unknown symbol
    pc 0x00007ff882e8c031 fp 0x00007ff8880fe6a0 Unknown symbol
    pc 0x00007ff884cb123c fp 0x00007ff8880fe6d8 Unknown symbol
    pc 0x00007ff8831ffdd0 fp 0x00007ff8880fe730 Unknown symbol
    pc 0x00007ff8831ea8db fp 0x00007ff8880fe770 Unknown symbol
    pc 0x00007ff8831ea480 fp 0x00007ff8880fe7b0 Unknown symbol
    pc 0x00007ff8831ea350 fp 0x00007ff8880fe7f0 Unknown symbol
    pc 0x00007ff8831ea12f fp 0x00007ff8880fe830 Unknown symbol
    pc 0x00007ff884cad7e2 fp 0x00007ff8880fe858 Unknown symbol
    pc 0x00007ff88860173a fp 0x00007ff8880fe8c8 Unknown symbol
    pc 0x000055b14d71fe75 fp 0x00007ff8880fe970 dart::DartEntry::InvokeFunction(dart::Function const&, dart::Array const&, dart::Array const&, unsigned long)
    pc 0x000055b14d723161 fp 0x00007ff8880fe9e0 dart::DartLibraryCalls::HandleMessage(dart::Object const&, dart::Instance const&)
    pc 0x000055b14d7758f9 fp 0x00007ff8880febe0 dart::IsolateMessageHandler::HandleMessage(std::__2::unique_ptr<dart::Message, std::__2::default_delete<dart::Message> >)
    pc 0x000055b14d7b969e fp 0x00007ff8880fec70 dart::MessageHandler::HandleMessages(dart::MonitorLocker*, bool, bool)
    pc 0x000055b14d7ba5e6 fp 0x00007ff8880fecf0 dart::MessageHandler::TaskCallback()
    pc 0x000055b14d962f4c fp 0x00007ff8880fed40 dart::ThreadPool::Worker::Loop()
    pc 0x000055b14d962a34 fp 0x00007ff8880fed90 dart::ThreadPool::Worker::Main(unsigned long)
    pc 0x000055b14d89e8d5 fp 0x00007ff8880feed0 out/DebugX64/dart+0x1aee8d5
  -- End of DumpStackTrace
  --- Re-run this test:
  python tools/test.py -n dartk-linux-debug-x64 service/async_generator_breakpoint_test/1
```

Crash in release mode:
```
  /===============================================================================\
  | service/async_generator_breakpoint_test/1 failed again (Crash, expected Pass) |
  \===============================================================================/
  --- Command "vm" (took 07.000417s):
  DART_CONFIGURATION=ReleaseX64 out/ReleaseX64/dart --enable_asserts --verbose-debug --stacktrace-every=55 --stress-async-stacks --ignore-unrecognized-flags --packages=/b/s/w/ir/.packages /b/s/w/ir/runtime/observatory/tests/service/async_generator_breakpoint_test.dart
  exit code:
  -6
  stdout:
  ** Launching SERVICE_TEST_TESTEE=true out/ReleaseX64/dart --disable-service-auth-codes --profiler --enable_asserts --verbose-debug --stacktrace-every=55 --stress-async-stacks --ignore-unrecognized-flags --packages=/b/s/w/ir/.packages --enable-vm-service:0 /b/s/w/ir/runtime/observatory/tests/service/async_generator_breakpoint_test.dart
  >testee>out> Observatory listening on http://127.0.0.1:37187/
  ** Signaled to run test queries on http://127.0.0.1:37187/
  >testee>out>
  Loading VM...
  >testee>out> testee waiting
  stderr:
  ===== CRASH =====
  si_signo=Segmentation fault(11), si_code=1, si_addr=0x7fa13c180000
  version=2.6.0-edge.1f9bf308b23e7bd18589dfd80295a90905ea07b8 (Thu Aug 29 15:49:59 2019 +0000) on "linux_x64"
  thread=8404, isolate=main(0x557104ede100)
    pc 0x0000557102f03743 fp 0x00007fa13b03d850 dart::PcDescriptors::DecodeInteger(long*) const
    pc 0x0000557102f0e408 fp 0x00007fa13b03d8c0 dart::Code::GetPcForDeoptId(long, dart::RawPcDescriptors::Kind) const
    pc 0x0000557102fd6a6b fp 0x00007fa13b03d900 dart::InlinedFunctionsIterator::InlinedFunctionsIterator(dart::Code const&, unsigned long)
    pc 0x0000557102e3a9c8 fp 0x00007fa13b03db10 dart::Debugger::CollectAwaiterReturnStackTrace()
    pc 0x0000557102faee02 fp 0x00007fa13b03dd20 dart::DRT_StackOverflow(dart::NativeArguments)
    pc 0x00007fa13b041108 fp 0x00007fa13b03dd60 Unknown symbol
    pc 0x00007fa135ff6f10 fp 0x00007fa13b03dd80 Unknown symbol
    pc 0x00007fa135ff66a8 fp 0x00007fa13b03ddf8 Unknown symbol
    pc 0x00007fa135ff62b1 fp 0x00007fa13b03de68 Unknown symbol
    pc 0x00007fa12f0d1c17 fp 0x00007fa13b03def8 Unknown symbol
    pc 0x00007fa12f10f0fc fp 0x00007fa13b03df60 Unknown symbol
    pc 0x00007fa12f10ee40 fp 0x00007fa13b03dfb0 Unknown symbol
    pc 0x00007fa12f1008d5 fp 0x00007fa13b03e030 Unknown symbol
    pc 0x00007fa12f1006b9 fp 0x00007fa13b03e0a8 Unknown symbol
    pc 0x00007fa12f43fcca fp 0x00007fa13b03e150 Unknown symbol
    pc 0x00007fa12f10c60c fp 0x00007fa13b03e1e0 Unknown symbol
    pc 0x00007fa12f10c0d6 fp 0x00007fa13b03e238 Unknown symbol
    pc 0x00007fa12f10b871 fp 0x00007fa13b03e278 Unknown symbol
    pc 0x00007fa12f10b463 fp 0x00007fa13b03e2b8 Unknown symbol
    pc 0x00007fa12f10b0c0 fp 0x00007fa13b03e2f8 Unknown symbol
    pc 0x00007fa12f10ac03 fp 0x00007fa13b03e338 Unknown symbol
    pc 0x00007fa12f10ca66 fp 0x00007fa13b03e388 Unknown symbol
    pc 0x00007fa12f10c860 fp 0x00007fa13b03e3d8 Unknown symbol
    pc 0x00007fa12f100953 fp 0x00007fa13b03e458 Unknown symbol
    pc 0x00007fa12f1006b9 fp 0x00007fa13b03e4d0 Unknown symbol
    pc 0x00007fa12f43fcca fp 0x00007fa13b03e578 Unknown symbol
    pc 0x00007fa12f10c60c fp 0x00007fa13b03e608 Unknown symbol
    pc 0x00007fa12f10c0d6 fp 0x00007fa13b03e660 Unknown symbol
    pc 0x00007fa12f10b871 fp 0x00007fa13b03e6a0 Unknown symbol
    pc 0x00007fa12f10b463 fp 0x00007fa13b03e6e0 Unknown symbol
    pc 0x00007fa12f10b0c0 fp 0x00007fa13b03e720 Unknown symbol
    pc 0x00007fa12f10ac03 fp 0x00007fa13b03e760 Unknown symbol
    pc 0x00007fa12f10a8d1 fp 0x00007fa13b03e7b0 Unknown symbol
    pc 0x00007fa1386b003c fp 0x00007fa13b03e7e8 Unknown symbol
    pc 0x00007fa12f43e8ef fp 0x00007fa13b03e840 Unknown symbol
    pc 0x00007fa12f42901b fp 0x00007fa13b03e880 Unknown symbol
    pc 0x00007fa12f428bc0 fp 0x00007fa13b03e8c0 Unknown symbol
    pc 0x00007fa12f428a90 fp 0x00007fa13b03e900 Unknown symbol
    pc 0x00007fa12f42886f fp 0x00007fa13b03e940 Unknown symbol
    pc 0x00007fa1386ac982 fp 0x00007fa13b03e968 Unknown symbol
    pc 0x00007fa13b04166c fp 0x00007fa13b03e9d8 Unknown symbol
    pc 0x0000557102e33fd6 fp 0x00007fa13b03ea80 dart::DartEntry::InvokeFunction(dart::Function const&, dart::Array const&, dart::Array const&, unsigned long)
    pc 0x0000557102e37486 fp 0x00007fa13b03eae0 dart::DartLibraryCalls::HandleMessage(dart::Object const&, dart::Instance const&)
    pc 0x0000557102e750a2 fp 0x00007fa13b03ece0 dart::IsolateMessageHandler::HandleMessage(std::__2::unique_ptr<dart::Message, std::__2::default_delete<dart::Message> >)
    pc 0x0000557102ea913d fp 0x00007fa13b03ed50 dart::MessageHandler::HandleMessages(dart::MonitorLocker*, bool, bool)
    pc 0x0000557102ea9986 fp 0x00007fa13b03edb0 dart::MessageHandler::TaskCallback()
    pc 0x0000557102fe71d7 fp 0x00007fa13b03ede0 dart::ThreadPool::Worker::Loop()
    pc 0x0000557102fe7075 fp 0x00007fa13b03ee20 dart::ThreadPool::Worker::Main(unsigned long)
    pc 0x0000557102f4e0e9 fp 0x00007fa13b03eed0 out/ReleaseX64/dart+0x190d0e9
  -- End of DumpStackTrace
  --- Re-run this test:
  python tools/test.py -n dartk-checked-linux-release-x64 service/async_generator_breakpoint_test/1
```

Can be reproduced on linux-debug-x64 with the following command:
```
tools/test.py --repeat 100 -n dartk-linux-debug-x64 service/async_generator_breakpoint_test/1
```

/cc @sjindel-google

**Comment by sjindel-google** [comment]
Adding FFI label since it probably affects FFI as well.

**Comment by crelier** [comment]
I have a fix for this one, but let me try to fix #38110 at the same time.
